sql/core/src/test/scala/org/apache/spark/sql/DataFrameRangeSuite.scala

```diff
@@ -153,23 +153,17 @@ class DataFrameRangeSuite extends QueryTest with SharedSQLContext with Eventuall
 
   test("Cancelling stage in a query with Range.") {
     val listener = new SparkListener {
-      override def onJobStart(jobStart: SparkListenerJobStart): Unit = {
-        eventually(timeout(10.seconds), interval(1.millis)) {
-          assert(DataFrameRangeSuite.stageToKill > 0)
-        }
-        sparkContext.cancelStage(DataFrameRangeSuite.stageToKill)
+      override def onTaskStart(taskStart: SparkListenerTaskStart): Unit = {
+        sparkContext.cancelStage(taskStart.stageId)
       }
     }
 
     sparkContext.addSparkListener(listener)
     for (codegen <- Seq(true, false)) {
       withSQLConf(SQLConf.WHOLESTAGE_CODEGEN_ENABLED.key -> codegen.toString()) {
-        DataFrameRangeSuite.stageToKill = -1
         val ex = intercept[SparkException] {
-          spark.range(0, 100000000000L, 1, 1).map { x =>
-            DataFrameRangeSuite.stageToKill = TaskContext.get().stageId()
-            x
-          }.toDF("id").agg(sum("id")).collect()
+          spark.range(0, 100000000000L, 1, 1)
```

Comment:
This is ok-ish but this kind of test is still racy. There's no guarantee the job won't finish before the events are posted to the bus, processed by the listener, and the stage is cancelled. The large count is just an attempt to make that less likely.

You could use a `CountDownLatch` for that - wait for it in the task (so that the task start event is fired), and signal it in the listener.

Reply:
+1

Reply:
Yeah, after extensive testing and tryout I was thinking about this as well but hesitating to stop the executor thread any kind of way. The reasons behind:
- Executor code can be synchronized only with object variables, otherwise `NotSerializableException` comes. As a result of introducing object `CountDownLatch ` similar race will appear just like `DataFrameRangeSuite.stageToKill` was overwritten.
- `spark.range(0, 100000000000L, 1, 1)` takes ~1-2 minutes to get calculated which is quite a big time window not to be flaky this way.
- Minor concern(rather note) if no timeout syncronization way used the test code get stuck in `collect` and never reaches `waitUntilEmpty` and `eventually...`.

All in all not recommend to do it because it could end up in 5 hours timeout.

Reply:
You could use `latch.await(timeout)` instead of just `latch.await()`, and throw an exception if it times out. That would avoid the test blocking indefinitely, and would add an explicit wait instead of using a large count to emulate it.

Reply:
How to overcome the first bullet point?

Reply:
You'd probably need to use the object field for that; which would be fine since you'd just be using it, not trying to overwrite it at any point.

But the current code is probably ok too. It's unlikely it will actually finish before the other threads have had a chance to run. I just wish that was a little bit more explicit in the code.